### PR TITLE
fs: extracts syscallfs.ReaderAtOffset for WASI and gojs

### PR DESCRIPTION
**Pull request by codefromthecrypt**
This extracts a utility `syscallfs.ReaderAtOffset()` to allow WASI and gojs to re-use the same logic to implement `syscall.Pread`.

What's different than before is that if WASI passes multiple iovecs an emulated `ReaderAt` will seek to the read position on each call to `Read` vs once per loop. This was a design decision to keep the call sites compatible between files that implement ReaderAt and those that emulate them with Seeker (e.g. avoid the need for a read-scoped closer/ defer function). The main use case for emulation is `embed.file`, whose seek function is cheap, so there's little performance impact to this.

```
goos: darwin
goarch: amd64
pkg: github.com/tetratelabs/wazero/internal/syscallfs
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkReaderAtOffset
BenchmarkReaderAtOffset/os.DirFS_io.File
BenchmarkReaderAtOffset/os.DirFS_io.File-16         	  549439	      2196 ns/op
BenchmarkReaderAtOffset/os.DirFS_readerAtOffset
BenchmarkReaderAtOffset/os.DirFS_readerAtOffset-16  	  612918	      1985 ns/op
BenchmarkReaderAtOffset/embed.FS_embed.file
BenchmarkReaderAtOffset/embed.FS_embed.file-16      	10059037	       126.1 ns/op
BenchmarkReaderAtOffset/embed.FS_seekToOffsetReader
BenchmarkReaderAtOffset/embed.FS_seekToOffsetReader-16         	 7055109	       179.0 ns/op
```

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
cc @ncruces as this moves around the stuff you made, and thanks for that!

**Comment by ncruces** [comment]
I'll try to have a stab at convincing the Go folks to add `ReaderAt` to `embed`. There's no good reason to for it not too. But that's always at least a year off.

**Comment by codefromthecrypt** [comment]
> I'll try to have a stab at convincing the Go folks to add ReaderAt to embed. There's no good reason to for it not too. But that's always at least a year off.

Thanks, I scratched my head looking at it, too, as it seems as easy as mapfs. You are right though, there is a socialization, review and landing concern. Thanks for offering to take that on which is much more effort than the code and tests!

File: internal/syscallfs/syscallfs_test.go [hunk 1]
```diff
@@ -257,3 +260,100 @@ func (f *testFSAdapter) Open(name string) (fs.File, error) {
 func requireErrno(t *testing.T, expected syscall.Errno, actual error) {
 	require.True(t, errors.Is(actual, expected), "expected %v, but was %v", expected, actual)
 }
+
+var (
+	//go:embed testdata/wazero.txt
+	readerAtFS   embed.FS
+	readerAtFile = "wazero.txt"
```
Comment (mathetake) [inline internal/syscallfs/syscallfs_test.go:267]: maybe better add empty file as well so that we won't crash when reading on the empty file?

**Review by mathetake — APPROVED** [review]

File: internal/syscallfs/syscallfs_test.go (hunk 1 above)
Comment (codefromthecrypt) [reply internal/syscallfs/syscallfs_test.go:267]: sure!

**Comment by ncruces** [comment]
Got it merged, should be in 1.21:
https://github.com/golang/go/issues/57803#event-8989467802

**Pull request merged**